### PR TITLE
Rewrite FP doc introduction

### LINKED ISSUE #1423: Functional Programming

### PR DESCRIPTION
**Pull request by cuihtlauac**
Rewrite goes up to line 333.

### CODE REVIEW

File: data/tutorials/lg_01_functional_programming.md [hunk 1]
```diff
@@ -11,7 +11,7 @@ date: 2021-08-06T17:16:00-00:00
 
 ## What Is Functional Programming?
 
-At its core, and from its inception, around 1985, OCaml is a [functional
+At its core, and from its origin, around 1985, OCaml is a [functional
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md]: Do you like “origin” better than “inception”? How about “from its debut” ?

At its core, and from its debut around 1985, OCaml is a ….

**Review by christinerose — APPROVED** [review]

File: data/tutorials/lg_01_functional_programming.md (hunk 1 above)
Comment (cuihtlauac) [reply data/tutorials/lg_01_functional_programming.md]: That's even better! Thanks a lot

**Review by SaySayo — APPROVED** [review]

**Review by christinerose — APPROVED** [review]

File: data/tutorials/lg_01_functional_programming.md [hunk 2]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
 
 The basic, and not very enlightening, distinction in **functional programming**
-is that **functions** are first-class citizens.
+is that **functions** are called **first-class citizens**. This means a function
+is just a value. The functions `sqrt`, `print_int` or `String.length` are values
+just like: `42`, `3.14` or `"haha"` are. In particular:
+
+1. Some functions can be written without giving them a name, like string
+   literals or arithmetic expressions
+1. Functions can be stored in data such as lists, arrays, trees, hash-tables and
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:102]:
```suggestion
1. Functions can be stored in data such as lists, arrays, trees, hash-tables, and
```

File: data/tutorials/lg_01_functional_programming.md [hunk 3]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+   others.
+1. Functions can be passed as parameters to other functions
+1. Functions can return functions as results
+
+Together, the two last items form what is called **higher-order**. In OCaml, the
+terms “functional value”, “non-functional value” and “value” denotes,
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:108]:
```suggestion
terms “functional value,” “non-functional value,” and “value” denotes,
```

File: data/tutorials/lg_01_functional_programming.md [hunk 4]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+respectively: functions, anything but a function or any of them.
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:109]:
```suggestion
respectively: functions, anything but a function, or any of them.
```

File: data/tutorials/lg_01_functional_programming.md [hunk 5]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+
+There are only two limitations to functions as values. They can't be displayed and
+they can't be compared.
+
+**Note**: Remember that the OCaml top-level runs code when it sees the double
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:114]:
```suggestion
**Note**: Remember that the OCaml toplevel runs code when it sees the double
```

File: data/tutorials/lg_01_functional_programming.md [hunk 6]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+semicolon `;;`. Throughout this tutorial, executable code examples begin with
+the command prompt `#`, end in `;;`, and show the expected output. If it
+doesn't, it's showcasing the code pattern rather than an executable code block.
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:117]: NICE!

File: data/tutorials/lg_01_functional_programming.md [hunk 7]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+
+### Function Calls
+
+Calling a function in OCaml is done be writing the parameter at the right of the
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:121]:
```suggestion
Calling a function in OCaml is done be writing the parameter to the right of the
```

File: data/tutorials/lg_01_functional_programming.md [hunk 8]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+```ocaml
+# max 3.14 6.28;;
+- : float = 6.28
+```
+
+In many programming languages, to call a function, its arguments need to be
+surrounded by parentheses and separated with commas, if there are several of
+them. It's not required in OCaml where parenthesis are only used to force the
+association of expressions, like in `3 * (x + 5)`. This is inherited from the
+&lambda;-calculus but it's also the same as in command lines such as `cp foo
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:138]:
```suggestion
&lambda;-calculus, but it's also the same as in command lines such as `cp foo
```

File: data/tutorials/lg_01_functional_programming.md [hunk 9]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+because that's what they are. Some even pronounce the above: “lambda x, three x plus
+5”.
+
+Anonymous functions expecting several arguments are written like this:
+```ocaml
+#(fun h r -> 1.04719755119659763 *. r *. r *. h);;
+ - : float -> float -> float = <fun>
+```
+
+### Function definitions
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:170]:
```suggestion
### Function Definitions
```

File: data/tutorials/lg_01_functional_programming.md [hunk 10]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+```
+Here the value `6.28318530717958623` is given the name `tau`. Defininig
+functions is the same and shows functions are just expressions ()
+```ocaml
+let g = fun x -> 3 * (x + 5);;
+val g : int -> int = <fun>
+```
+
+In some languages, there is a distinction between expressions and statements.
+Expressions have values like `true`, `2 * 3.14` or `"hello %s!\n"`, statements
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:227]:
```suggestion
Expressions have values like `true`, `2 * 3.14`, or `"hello %s!\n"`; statements
```

File: data/tutorials/lg_01_functional_programming.md [hunk 11]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+aren't values. For instance, in C, `if (a) b = 7; else b = 12;` is a statement,
+not an expression. In such a context, the values are the first-class citizens
+and statements are the second-class citizens. In OCaml, everything is an
+expression, there are no statements. For instance, in OCaml, it is possible to
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:231]:
```suggestion
expression; there are no statements. For instance, in OCaml, it is possible to
```

File: data/tutorials/lg_01_functional_programming.md [hunk 12]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+```
+Here is a list of functions:
+```ocaml
+#[sin; cos; fun x -> x *. x];;
+- : (float -> float) list = [<fun>; <fun>; <fun>]
+```
+
+What makes functions as values really important is the capability to have
+functions as parameter or result of other functions. That's what is called
+higher-order functions. This term has an arcane origin, it's better to take it
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:250]:
```suggestion
higher-order functions. This term has an arcane origin, so it's better to take it
```

File: data/tutorials/lg_01_functional_programming.md [hunk 13]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+let cone_volume = fun h r -> 1.04719755119659763 *. r *. r *. h;;
+val cone_volume : float -> float -> float = <fun>
+```
+
+There's a fourth way to write the same thing:
+```ocaml
+# let cone_volume = fun h -> fun r -> 1.04719755119659763 *. r *. r *. h;;
+val cone_volume : float -> float -> float = <fun>
+```
+Here the body of the function isn't a &lambda; binding two names, it's a
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:272]:
```suggestion
Here the body of the function isn't a &lambda; binding two names; it's a
```

File: data/tutorials/lg_01_functional_programming.md [hunk 14]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+&lambda; nested inside another. The top-most binds `h` whilst the inner one
+binds `r`. It means `cone_volume` is defined as a function which returns a
+function, _ergo_ a higher-order function. This last form should be considered
+the raw one, the others are sugar-coated versions of the last one.
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:276]:
```suggestion
the raw one. The others are sugar-coated versions of the last one.
```

File: data/tutorials/lg_01_functional_programming.md [hunk 15]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+
+Concerning function application, this means the raw form is `(cone_volume 3.0)
+5.0`, the form `cone_volume 3.0 5.0` is the sugar-coated equivalent.
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:279]:
```suggestion
5.0`, and the form `cone_volume 3.0 5.0` is the sugar-coated equivalent.
```

File: data/tutorials/lg_01_functional_programming.md [hunk 16]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+
+A short way to remember this is to memorize: “all functions are unary”. Some
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:281]:
```suggestion
A short way to remember this is to memorize: “all functions are unary.” Some
```

File: data/tutorials/lg_01_functional_programming.md [hunk 17]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+functions may look as if they take several parameters, but it's a trick, they
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:282]:
```suggestion
functions may look as if they take several parameters, but it's a trick! They actually 
```

File: data/tutorials/lg_01_functional_programming.md [hunk 18]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+take a single argument and return a function with one parameter less. We'll come
+back to this later.
+
+The second side of higher-order functions is passing functions to functions.
+There are plenty of them in the standard library, here are a couple of them:
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:287]:
```suggestion
There are plenty of them in the standard library. Here are a couple of them:
```

File: data/tutorials/lg_01_functional_programming.md [hunk 19]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+val list_double : int list -> int list = <fun>
+# list_double [1; 2; 3];;
 - : int list = [2; 4; 6]
 ```
 
-In this example, we first defined a nested function called `double`
-that took the argument `x` and returned `x * 2`. Then `map` calls
-`double` on each element of the given list (`[1; 2; 3]`) to produce the
-result: a list with each number doubled.
+Now, the definition of `double` is local, as previously, but it is also nested
```
Comment (christinerose) [inline data/tutorials/lg_01_functional_programming.md:330]:
```suggestion
Now, the definition of `double` is local, as previously, but it is also nested,
```

**Review by christinerose — CHANGES_REQUESTED** [review]
Minor grammar and formatting suggestions :)

File: data/tutorials/lg_01_functional_programming.md [hunk 20]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+- Domains-based support for multi-core hardware architectures
+- Distributed computing using effect-handlers
+
+Don't worry if those terms aren't making sense, yet. They are detailed in others
+documents. This is just a list of keywords.
+
+It is important to remember that all the features of OCaml are either built on
+top of functional programming or meant to support it. Most mainstream languages
+derive from the imperative programming paradigm, to which either structured or
+object-oriented programming where added. The feature set of OCaml, as sketched
```
Comment (gpetiot) [inline data/tutorials/lg_01_functional_programming.md]: "were"

File: data/tutorials/lg_01_functional_programming.md [hunk 21]
```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
...
+earlier, was pretty unique for many years. Nowadays, several languages have many
+of the features found in OCaml, but either their conception or their evolution
+goes in the opposite direction. OCaml started from functional programming while
+many others are going towards functional programming. This is not the place to
+detail the differences these paths entail. Let's just say OCaml developers and
+practitioners believe the outstanding level of reliability, performance and
```
Comment (gpetiot) [inline data/tutorials/lg_01_functional_programming.md]: Could we just use "users"? "practitioners" has a medicine tone (maybe that's on purpose)

**Comment by sabine** [comment]
Closing since "Higher Order Functions" will most likely replace this document entirely.

**Pull request closed without merging**